pkg/scheduler/factory.go

```diff
@@ -480,40 +473,9 @@ func MakeDefaultErrorFunc(client clientset.Interface, podQueue internalqueue.Sch
 			klog.Errorf("Error scheduling %v/%v: %v; retrying", pod.Namespace, pod.Name, err)
 		}
 
-		podSchedulingCycle := podQueue.SchedulingCycle()
-		// Retry asynchronously.
-		// Note that this is extremely rudimentary and we need a more real error handling path.
-		go func() {
-			defer utilruntime.HandleCrash()
-			podID := types.NamespacedName{
-				Namespace: pod.Namespace,
-				Name:      pod.Name,
-			}
-
-			// Get the pod again; it may have changed/been scheduled already.
-			getBackoff := initialGetBackoff
-			for {
-				pod, err := client.CoreV1().Pods(podID.Namespace).Get(context.TODO(), podID.Name, metav1.GetOptions{})
-				if err == nil {
-					if len(pod.Spec.NodeName) == 0 {
-						podInfo.Pod = pod
-						if err := podQueue.AddUnschedulableIfNotPresent(podInfo, podSchedulingCycle); err != nil {
-							klog.Error(err)
-						}
-					}
-					break
-				}
-				if apierrors.IsNotFound(err) {
-					klog.Warningf("A pod %v no longer exists", podID)
-					return
-				}
-				klog.Errorf("Error getting pod %v for retry: %v; retrying...", podID, err)
-				if getBackoff = getBackoff * 2; getBackoff > maximalGetBackoff {
-					getBackoff = maximalGetBackoff
-				}
-				time.Sleep(getBackoff)
-			}
-		}()
+		if err := podQueue.AddUnschedulableIfNotPresent(podInfo, podQueue.SchedulingCycle()); err != nil {
+			klog.Error(err)
+		}
```

Comment:
note that if the failure happens in the binding go routine, the pod here will get placed in the queue, and the main thread may pick it up before the update to the pod condition happens, and so we miss that too.

Reply:
> what if the pod gets updated while it was being processed?

This PR doesn't change how it's processed before, say:

- If the being processed Pod is schedulable, no matter the updated Pod is schedulable or not, the Pod finished its binding cycle and get its `nodeName` set.
- If the being processed Pod is unschedulable, putting it back to schedulingQ may conflict with the updated version, but it's fault tolerant. Next time when it's popped out, we have the latest copy.

> aren't we going to miss those updates since the pod is not in the queue at all?

This PR only put extra logic on delete case, so no changes on updated Pod.

> perhaps we can read the pod back from informer cache.

That's another approach we can exercise, then we don't need the `processingPods` map.

Reply:
>If the being processed Pod is unschedulable, putting it back to schedulingQ may conflict with the updated version, but it's fault tolerant. Next time when it's popped out, we have the latest copy.

you may miss the update because the error function could have been called from the bind goroutine as I mentioned in https://github.com/kubernetes/kubernetes/pull/90660/files?file-filters%5B%5D=.go#r421165418, here is how this could happen in your original version:
1) pod is added to the queue
2) main thread pops the pod from the queue
3) update pod condition happens, but the event is missed because the pod is now not in the queue



> This PR only put extra logic on delete case, so no changes on updated Pod.

I know, but by removing the pod fetch from the api server, the updates that could have happened to the pod while being processed would be lost. This is not an issue now that we are fetching the pod from the informer cache given that we use it to update the podinfo passed to AddUnschedulableIfNotPresent as the original code did: https://github.com/kubernetes/kubernetes/pull/90660/files?file-filters%5B%5D=.go#diff-ec926856bfdbbcaf5e9dfce784909d51L499 .

Reply:
> 3. update pod condition happens, but the event is missed because the pod is now not in the queue

I don't quite follow how "the event is missed". In this case, isn't the updated Pod added to activeQ?

https://github.com/kubernetes/kubernetes/blob/be02fe578588592ee67811ed809eef0a4495537a/pkg/scheduler/internal/queue/scheduling_queue.go#L451-L452

Reply:
Nvm, I understand it now. You're referring that when the stale version of pod gets added back to scheduler queue, it would "overwrite" the fresh version, so the fresh version gets lost, right?

Reply:
> I don't quite follow how "the event is missed". In this case, isn't the updated Pod added to activeQ?

Yes, but there is a race, imagine that the main thread poped the pod from the active queue and started to process it again before the update event arrives.